File: LeetCode/Seattle/Tree/SymmetricTree.cs
using System;
using System.Collections.Generic;
using System.Text;
using Seattle.DataStructure;
using Xunit;

namespace Seattle.TreeAlgorithms
{
    /// <summary>
    /// https://leetcode.com/problems/symmetric-tree/.
    /// </summary>
    public class SymmetricTree
    {
        public bool IsSymmetric(TreeNode root)
        {
            return IsMirror(root, root);
        }

        private bool IsMirror(TreeNode r, TreeNode l)
        {
            if (r == null && l == null) return true;
            if (r == null || l == null) return false;
            return (r.val == l.val)
                   && IsMirror(r.right, l.left)
                   && IsMirror(r.left, l.right);
        }
    }

    [Collection(SeattleCollectionFixture.SeattleCollectionFixtureName)]
    public class SymmetricTreeTests
    {
        public SymmetricTreeTests(SeattleTestsFixture seattleTestsFixture)
        {
        }

        [Fact]
        public void Test()
        {
            var algorithm = new SymmetricTree();

            //     1
            //    / \
            //   2   2
            //  / \ / \
            // 3  4 4  3
            var t2 = new Tree();
            int[] arr = {1, 2, 2, 3, 4, 4, 3};
            t2.root = t2.insertLevelOrder(arr, t2.root, 0);
            t2.inOrder(t2.root);

            var res = algorithm.IsSymmetric(t2.root);
            Assert.True(res);

            //      1
            //    /  \
            //   2    2
            //  / \  / \
            // 5  3  7  3
            arr = new int[] { 1, 2, 2, 6, 3, 7, 3 };
            t2.root = t2.insertLevelOrder(arr, t2.root, 0);
            t2.inOrder(t2.root);

            res = algorithm.IsSymmetric(t2.root);
            Assert.False(res);
        }
    }
}

File: py/Seattle/pythonProj/Tree/SymmetricTree.py
from typing import List
from TreeNode import TreeNode

class Solution:
    # https://leetcode.com/problems/symmetric-tree/
    def isSymmetric(self, r: TreeNode) -> bool:
        return self.isMirror(r, r)

    def isMirror(self, l: TreeNode, r: TreeNode) -> bool:
        if l is None and r is None: return True
        if l is None or r is None: return False
        if l.val == r.val and self.isMirror(l.left, r.right) and self.isMirror(l.right, r.left):
            return True
        else:
            return False


if __name__ == '__main__':
    sln = Solution()

    t2 = TreeNode()
    arr = [1, 2, 2, 3, 4, 4, 3]
    t2.root = t2.insert_level_order(arr, t2.root, 0)
    t2.in_order(t2.root);
    print(sln.isSymmetric(t2.root))

    arr = [1, 2, 2, 6, 3, 7, 3]
    t2.root = t2.insert_level_order(arr, t2.root, 0)
    t2.in_order(t2.root)
    print(sln.isSymmetric(t2.root))